### PR TITLE
build(snap): update snap packaging

### PR DESCRIPTION
**Pull request by siggiskulason**
- Update snap base to core 20
- Update documentation
- Update metadata and versioning
- Add support for "snap set" of device-specific settings
- Update install hook
- Other minor updates

Signed-off-by: Siggi Skulason <siggi.skulason@canonical.com>

Fix https://github.com/edgexfoundry/app-rfid-llrp-inventory/issues/58

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-rfid-llrp-inventory/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [X] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?) N/'A
- [X] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?) N/A
  <link to docs PR>

## Testing Instructions

It's not possible to fully test this without a reader. What I did, for reference, is:

1. Remove and reinstall edgexfoundry and the LLRP-RFID snaps

```
sudo snap remove edgexfoundry
sudo snap remove edgex-device-rfid-llrp
sudo snap remove edgex-app-rfid-llrp-inventory
sudo snap install edgexfoundry --edge
```

2. Run the device-rfid-llrp snap from https://github.com/edgexfoundry/device-rfid-llrp-go/pull/61. Connect the interfaces and run the scan to get the subnet information.

```
snapcraft clean --use-lxd
snapcraft --use-lxd
sudo snap install edgex-device-rfid-llrp_1.0.1-dev.7_amd64.snap --dangerous
sudo snap connect edgexfoundry:edgex-secretstore-token edgex-device-rfid-llrp:edgex-secretstore-token
sudo snap start edgex-device-rfid-llrp.device-rfid-llrp 
```

```
sudo snap connect edgex-device-rfid-llrp:network-control 
sudo snap connect edgex-device-rfid-llrp:network-observe 

CONSUL_TOKEN=$(sudo cat /var/snap/edgexfoundry/current/secrets/consul-acl-token/bootstrap_token.json | jq ".SecretID" | tr -d '"') 
echo $CONSUL_TOKEN
sudo edgex-device-rfid-llrp.auto-configure $CONSUL_TOKEN
```

3. Wait for the RFID reader to be detected (takes a few seconds). Confirm using `edgex-cli` that the device was added:

```
$ edgex-cli device list
Name                Description       ServiceName       ProfileName          Labels                           AutoEvents
SpeedwayR-10-EE-EF  LLRP RFID Reader  device-rfid-llrp  LLRP-Impinj-Profile  [RFID LLRP Impinj SpeedwayR420]  []
```

3. Using the name of the device, update the configuration.toml file for this inventory snap, adding the required alias:

```
[AppCustom.Aliases]
SpeedwayR-10-EE-EF_1="office"
```
Note, that as per the Readme, the key is the device name plus "_1" for the 1st antenna.

4. Build, install and run the snap

```
snapcraft --use-lxd
sudo snap install edgex-app-rfid-llrp-inventory_1.0.1-dev.6_amd64.snap --dangerous
sudo snap connect edgexfoundry:edgex-secretstore-token edgex-app-rfid-llrp-inventory:edgex-secretstore-token
sudo snap start edgex-app-rfid-llrp-inventory.app-rfid-llrp-inventory 
```

5. Use the Consul UI to double-check that the AppCustom/Aliases value is correctly set

6. Run a scan for a few seconds and check the results

```
curl -o- -X POST localhost:59711/api/v2/command/reading/start
sleep 5s
curl -o- -X POST localhost:59711/api/v2/command/reading/stop
sleep 3s
curl -o- localhost:59711/api/v2/inventory/snapshot
```
Result:

```
[{"epc":"e200001b690901820340ee3f","tid":"","location":{"device_name":"SpeedwayR-10-EE-EF","antenna_id":1},"location_alias":"office","last_read":1645042805425,"last_arrived":1645042800663,"last_departed":0,"state":"Present","stats_map":{"SpeedwayR-10-EE-EF_1":{"last_read":1645042805425,"mean_rssi":-45.425}}},{"epc":"e200001b690901820330ef6c","tid":"","location":{"device_name":"SpeedwayR-10-EE-EF","antenna_id":1},"location_alias":"office","last_read":1645042805425,"last_arrived":1645042800663,"last_departed":0,"state":"Present","stats_map":{"SpeedwayR-10-EE-EF_1":{"last_read":1645042805425,"mean_rssi":-33.525}}}]
```

### CODE REVIEW

**Review by lenny-intel — COMMENTED** [review]
LGTM, i.e. nothing changed outside `snap/` ;-) Leaving it to other snap expected to approve.

**Comment by siggiskulason** [comment]
@farshidtz - please review. Thanks!

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -1,27 +1,20 @@
...
 
 grade: stable
 confinement: strict
 
-# delhi: 0, edinburgh: 1, fuji: 2, geneva: 3, ireland: 4, jakarta: 5
+# 1: edinburgh
+# 2: fuji
+# 3: geneva, hanoi
+# 4: ireland, 
+#5:  jakarta
```
Comment (farshidtz) [inline snap/snapcraft.yaml:17]:
```suggestion
# 5:  jakarta
```

File: snap/README.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+## Service Environment Configuration Overrides
+
+### [Service]
+|snap set|configuration.yaml setting|
+|---|---|
+|env.service.boot-timeout|Service.BootTimeout|
+|env.service.health-check-interval|Service.HealthCheckInterval|
+|env.service.host|Service.Host|
+|env.service.server-bind-addr|Service.ServerBindAddr|
+|env.service.port|Service.Port|
```
Comment (farshidtz) [inline snap/README.md:124]: This doesn't work:

```
$ sudo snap set edgex-app-rfid-llrp-inventory env.service.port=55555
$ sudo snap restart edgex-app-rfid-llrp-inventory
```

Logs:
```
$ sudo snap logs -f -n 100 edgex-app-rfid-llrp-inventory.app-rfid-llrp-inventory | grep localhost
...
2022-02-17T18:18:07+01:00 edgex-app-rfid-llrp-inventory.app-rfid-llrp-inventory[62516]: level=INFO ts=2022-02-17T17:18:07.503249243Z app=app-rfid-llrp-inventory source=server.go:156 msg="Starting HTTP Web Server on address localhost:59711"
...
2022-02-17T18:28:57+01:00 edgex-app-rfid-llrp-inventory.app-rfid-llrp-inventory[65781]: level=INFO ts=2022-02-17T17:28:57.713082759Z app=app-rfid-llrp-inventory source=server.go:156 msg="Starting HTTP Web Server on address localhost:59711"
```

Tried `server-bind-addr` too; it had no effect.

**Review by farshidtz — CHANGES_REQUESTED** [review]
@siggiskulason, thanks a lot for the updated. I've made some inline comments, mostly non-functional issues. But there is a problem with env overrides set via snap options.

File: snap/README.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+|env.service.protocol|Service.Protocol|
```
[the remark this answers is not in the document]
Comment (siggiskulason) [reply snap/README.md]: removed from README

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -46,7 +46,17 @@ apps:
     plugs: [network, network-bind]
```
[the remark this answers is not in the document]
Comment (siggiskulason) [reply snap/snapcraft.yaml:44]: removed

**Comment by siggiskulason** [comment]
@farshidtz - all requested issues have been resolved. Please re-review. thanks!

**Comment by siggiskulason** [comment]
force-pushed to fix author/signoff issues.

**Review by farshidtz — APPROVED** [review]
Thanks for the updates. I tested the packaging and runtime and it works as expected.

**Comment by siggiskulason** [comment]
Thanks @farshidtz . - @lenny-intel could you please approve/merge this PR? (Farshid is not on the committer list)

**Review by lenny-intel — APPROVED** [review]

**Pull request merged**